Namespace package (#26)
MatterMiners/cobald: docs/source/custom/package.rst

=================================
Using and Distributing Extensions
=================================

> Extensions for :py:mod:`cobald` are regular Python code accessible to the interpreter.

Reviewer: ```suggestion
Extensions for :py:mod:`cobald` is regular Python code accessible to the interpreter.
```

For specific problems, extensions can be defined directly in a Python configuration file.

> General purpose and reusable code should be made available as a regular Python package.

Reviewer: Why is it important that it is a *regular* Python package. Are there any other options?

```suggestion
General purpose and reusable code should be made available as a Python package.
```

This ensures proper installation and dependency management.

Configuration Files
===================

Using Python :doc:`configuration files </source/daemon/config>` allows to define arbitrary
objects, functions and helpers.
This is ideal for minor modifications of existing objects and
experimental extensions.
Simply add new definitions to the configuration before using them:

.. code:: python3

    #/etc/cobald/my_demo.py
    from cobald.interface import Controller

    from cobald_demo.cpu_pool import CpuPool
    from cobald_demo.draw_line import DrawLineHook

    # custom Controller implementation
    class StaticController(Controller):
        """Controller that sets demand to a fixed value"""
        def __init__(self, target, demand):
            super().__init__(target)
            self.target.demand = demand

    # use custom Controller

>     pipeline = StaticController.s(demand=50) >> DrawLineHook.s() >> CpuPool(interval=1)

Reviewer: Please consider, if Pools should be initialised similar to the Controllers, e.g. `CpuPool.s(interval=1)`.

Author: I've opened a new ticket #32 to address this. IMO this is useful to have, but not as part of this pull request.

Configuration files are easy to use and modify, but impractical for reusable extensions.

Python Packages
===============

For generic extensions, Python packages simplify distribution and reuse.
Packages are individual `.py` files or folders containing several `.py` files;

> in addition, packages contain meta-data for dependency management and installation.

Reviewer: ```suggestion
in addition, packages contain metadata for dependency management and installation.
```

.. code:: python3

    # my_controller.py
    from cobald.interfaces import Controller

    class StaticController(Controller):
        def __init__(self, target, demand):
            super().__init__(target)
            self.target.demand = demand

> Packages can be temporarily mounted via ``PYTHONPATH`` or permanently installed.

Reviewer: Is it correct here to talk about *mounting*? I feel like *loading* might be better?

```suggestion
Packages can be temporarily loaded via ``PYTHONPATH`` or permanently installed.
```

Author: ```suggestion
Packages can be temporarily accessed via ``PYTHONPATH`` or permanently installed.
```

> Once available, packages can be imported and used in any configurations.

Reviewer: ```suggestion
Once available, packages can be imported and used in any configuration.
```

.. code:: python3

    #/etc/cobald/my_demo.py
    from my_controller import StaticController

    from cobald_demo.cpu_pool import CpuPool
    from cobald_demo.draw_line import DrawLineHook

    # use custom Controller from package

>     pipeline = StaticController.s(demand=50) >> DrawLineHook.s() >> CpuPool(interval=1)

Reviewer: Same as above.
